src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/ServiceDescriptionService.java

```diff
@@ -534,6 +581,71 @@ public ServiceDescriptionType refreshServiceDescription(Long id, boolean ignoreW
     }
 
+    /**
+     * Refresh REST service description
+     *
+     * @param serviceDescriptionType
+     * @return {@link ServiceDescriptionType}
+     * @throws WrongServiceDescriptionTypeException if service type is not REST
+     * @throws ServiceNotFoundException             if there is no service linked to service description
+     * @throws InvalidUrlException                  if url is invalid
+     */
+    private ServiceDescriptionType refreshRESTServiceDescription(ServiceDescriptionType serviceDescriptionType)
+            throws WrongServiceDescriptionTypeException, ServiceNotFoundException, InvalidUrlException {
+        verifyAuthority("REFRESH_REST");
+
+        if (!serviceDescriptionType.getType().equals(DescriptionType.REST)) {
+            throw new WrongServiceDescriptionTypeException("Expected description type REST");
+        }
+
+        validateUrl(serviceDescriptionType.getUrl());
+
+        serviceDescriptionType.setRefreshedDate(new Date());
+
+        return serviceDescriptionType;
+    }
+
+    /**
+     * Refresh OPENAPI3 ServiceDescription
+     *
+     * @param serviceDescriptionType
+     * @param ignoreWarnings
+     * @return {@link ServiceDescriptionType}
+     * @throws WrongServiceDescriptionTypeException if service type is not openapi3
+     * @throws ServiceNotFoundException             if there is no service linked to given service description
+     * @throws UnhandledWarningsException           if unhandled warnings are found and ignoreWarnings if false
+     * @throws OpenApiParser.ParsingException       if parsing openapi3 description fails
+     * @throws InvalidUrlException                  if url is invalid
+     */
+    private ServiceDescriptionType refreshOPENAPI3ServiceDescription(ServiceDescriptionType serviceDescriptionType,
+            boolean ignoreWarnings) throws WrongServiceDescriptionTypeException, ServiceNotFoundException,
+            UnhandledWarningsException, OpenApiParser.ParsingException, InvalidUrlException {
+
+        verifyAuthority("REFRESH_OPENAPI3");
+
+        if (!serviceDescriptionType.getType().equals(DescriptionType.OPENAPI3)) {
+            throw new WrongServiceDescriptionTypeException("Expected description type OPENAPI3");
+        }
+
+        if (serviceDescriptionType.getService().get(0) == null) {
+            throw new ServiceNotFoundException("Service not found from servicedescription with id "
+                    + serviceDescriptionType.getId());
+        }
+
+        validateUrl(serviceDescriptionType.getUrl());
+
+        serviceDescriptionType.setRefreshedDate(new Date());
+
+        parseOpenapi3ToServiceDescription(serviceDescriptionType.getUrl(),
+                serviceDescriptionType.getService().get(0).getServiceCode(),
+                ignoreWarnings,
+                serviceDescriptionType);
+
+        clientRepository.saveOrUpdateAndFlush(serviceDescriptionType.getClient());
+
+        return serviceDescriptionType;
```

Comment:
This handles warnings pretty differently from what `refreshWSDLServiceDescription` does. WSDL side gives warnings **if new services were added or removed**, this does not do it AFAIK?

- I think this is how old UI works too, right? Old UI does not warn about adding / removing openapi services either?
- is this by design or accidental?

Reply:
Yes, old UI doesn't tell if endpoints are changed (refresh doesn't have effect on services).

Reply:
Warnings will be implemented in a separate ticket https://jira.niis.org/browse/XRDDEV-918